### PR TITLE
4.x - Add type hints to database package

### PR DESCRIPTION
**Pull request by markstory**
I had to move the part lists into Query::traverse() to maintain interface consistentcy with ExpressionInterface. Moving part lists also required a more clunky implementation for SQLServer, but that driver already has a few clunky parts.

A few other types needed to become nullable as we use them in a nullable way in test cases. I was concerned that removing those tests would break real-world usage.

Refs #11935

### CODE REVIEW

File: src/Database/Log/QueryLogger.php [hunk 1]
```diff
@@ -45,7 +45,7 @@ public function log(LoggedQuery $query)
      * @param \Cake\Database\Log\LoggedQuery $query to be written in log
      * @return void
      */
-    protected function _log($query)
+    protected function _log(\Cake\Database\Log\LoggedQuery $query): void
```
Comment (ADmad) [inline src/Database/Log/QueryLogger.php]: Need to add "use" statement instead. Same for `_interpolate()` below.

File: src/Database/SqlserverCompiler.php [hunk 2]
```diff
@@ -62,7 +54,7 @@ class SqlserverCompiler extends QueryCompiler
      * @param \Cake\Database\ValueBinder $generator the placeholder generator to be used in expressions
      * @return string
      */
-    protected function _buildInsertPart($parts, $query, $generator)
+    protected function _buildInsertPart(array $parts, \Cake\Database\Query $query, \Cake\Database\ValueBinder $generator): string
```
Comment (ADmad) [inline src/Database/SqlserverCompiler.php]: FQCN need to be changed.

File: src/Database/Dialect/SqlserverDialectTrait.php [hunk 3]
```diff
@@ -335,32 +335,32 @@ public function schemaDialect(): BaseSchema
     /**
      * Returns a SQL snippet for creating a new transaction savepoint
      *
-     * @param string $name save point name
+     * @param string|int $name save point name
```
Comment (dereuromark) [inline src/Database/Dialect/SqlserverDialectTrait.php]: Why do we need to allow int here?

File: src/Datasource/QueryInterface.php [hunk 4]
```diff
@@ -159,7 +159,7 @@ public function count(): int;
      * $query->limit($query->newExpr()->add(['1 + 1'])); // LIMIT (1 + 1)
      * ```
      *
-     * @param int $num number of records to be returned
+     * @param int|mixed $num number of records to be returned
```
Comment (dereuromark) [inline src/Datasource/QueryInterface.php]: The mixed here sounds bad - as an addition it makes it pretty much worthless to expect and input type.

File: src/Database/Query.php [hunk 5]
```diff
@@ -277,12 +309,12 @@ public function sql(?ValueBinder $generator = null): string
...
      * @param callable $visitor A function or callable to be executed for each part
-     * @param array $parts The query clauses to traverse
      * @return $this
      */
-    public function traverse(callable $visitor, array $parts = [])
+    public function traverse(callable $visitor)
     {
-        $parts = $parts ?: array_keys($this->_parts);
+        $partList = "_{$this->type()}Parts";
+        $parts = $this->{$partList} ?? array_keys($this->_parts);
```
Comment (lorenzo) [inline src/Database/Query.php]: I would really like to keep the parts in the query compiler, unless you think this is a better idea.

One way to keep that design is by adding a traverseParts() function to the query, with the signature and implementation this function had before.

File: src/Datasource/QueryInterface.php (hunk 4 above)
Comment (markstory) [reply src/Datasource/QueryInterface.php]: Datasource doesn't contain a way to describe a QueryExpression which is what the database implementation can accept in addition to an int. Ither implementations may accept othe 'expression' style objects as well.

File: src/Database/Query.php (hunk 5 above)
Comment (markstory) [reply src/Database/Query.php]: I am not convinced what I have is great. Adding a `traverseParts` method to Query sounds good to me.

File: src/Database/Dialect/SqlserverDialectTrait.php (hunk 3 above)
Comment (markstory) [reply src/Database/Dialect/SqlserverDialectTrait.php]: Tests were failing as some of the db implementationa were paasing integers.

Comment (dereuromark) [reply src/Database/Dialect/SqlserverDialectTrait.php]: I see, so casting beforehand is not feasible here.

**Pull request merged**